### DISCUSSION TITLE
How can i get client's ip in middleware?

### DISCUSSION THREAD

**Discussion by YusufcanY** [nuxt/nuxt#17111] [Questions] [answered]
I just want to get clients ip and get country of that ip. I tried something like this:
```ts
export default defineNuxtRouteMiddleware(async (to, from) => {
  const ip = useRequestHeaders(['x-forwarded-for'])
  const { data } = await useFetch(`https://ipapi.co/${ip['x-forwarded-for']}/country`)
  if (data.value === 'US' || data.value === 'CA') {
    externalRedirect('https://blabla.com')
  }
})
```
Anyone can help?

**Comment by Diizzayy** [comment, marked as answer]
@YusufcanY You're on the right track with attempting to extract the `x-forwarded-for` header, though based on your setup (ie: if your service is behind a proxy), you'll need to ensure that the pertinent headers are indeed passed.

Another critical aspect to be aware of is that `useRequestHeaders` is a server side helper, hence when called on client side, It's result will be an empty object. Basically the IP address would only be retrievable on the server side ( assuming that the pertinent headers are indeed available as mentioned above ), you can also just use a check ( `if (process.server) `to only execute your ip logic on server.

If absolutely needed, a workaround to access the IP address on client-side ( security implications need to be taken into consideration ) would be to initiate it in a server plugin and propagate it to the client using [`useState`](https://v3.nuxtjs.org/guide/features/state-management)

`nuxt-app/plugins/ip.server.ts`
```ts
export default defineNuxtPlugin(() => {
  const ip = useState('ip', () => null)

  const event = useRequestEvent()

  // attempt to get IP
  ip.value = event.req.headers['x-real-ip'] || event.req.headers['x-forwarded-for'] || event.req.socket?.remoteAddress
})
```

Using the above, the IP should be accessible to the client-side

```ts
export default defineNuxtRouteMiddleware(async (to, from) => {
  const ip = useState('ip', () => null)

  if (ip.value) {
    const { data } = await useFetch(`https://ipapi.co/${ip.value}/country`)
    if (data.value === 'US' || data.value === 'CA') {
      externalRedirect('https://blabla.com')
    }
  }
})
```

**Reply by jboada** [reply]
So far, this is the only post I have seen about getting the client's ip address.
